services/debugger/cache/embeddedcache/embeddedCache.go

```
@@ -0,0 +1,192 @@
+	config.RegisterDurationConfigVariable(5, &e.Ticker, false, time.Minute, "LiveEvent.cache.GCTime")
+	config.RegisterDurationConfigVariable(15, &e.queryTimeout, false, time.Second, "LiveEvent.cache.queryTimeout")
+	config.RegisterIntConfigVariable(3, &e.retries, false, 1, "LiveEvent.cache.retries")
+}
+
+/*
+EmbeddedCache is an in-memory cache. Each key-value pair stored in this cache have a TTL and one goroutine removes the
+key-value pair form the cache which is older than TTL time.
+*/
+type EmbeddedCache[E any] struct {
+	Limiter      int
+	retries      int
+	path         string
+	Origin       string
+	done         chan struct{}
+	closed       chan struct{}
+	Ticker       time.Duration
+	queryTimeout time.Duration
+	CleanupFreq  time.Duration // TTL time on badgerDB
+	once         sync.Once
+	Db           *badger.DB
+	Logger       logger.Logger
+}
+
+type badgerLogger struct {
+	logger.Logger
+}
+
+func (badgerLogger) Errorf(format string, a ...interface{}) {
+	_, _ = fmt.Fprintf(os.Stderr, format, a...)
+}
+
+func (badgerLogger) Warningf(format string, a ...interface{}) {
+	_, _ = fmt.Fprintf(os.Stderr, format, a...)
+}
+
+// Update writes the entries into badger db with a TTL
+func (e *EmbeddedCache[E]) Update(key string, value E) {
+	e.Init()
+	txn := e.Db.NewTransaction(true)
+	res := e.Read(key)
+	if len(res) >= e.Limiter {
+		res = res[len(res)-e.Limiter+1:]
+	}
+	res = append(res, value)
+	data, err := json.Marshal(res)
+	if err != nil {
+		return
+	}
+	setBadgerEntry := func(ctx context.Context) error {
+		entry := badger.NewEntry([]byte(key), data).WithTTL(e.CleanupFreq)
+		if err := txn.SetEntry(entry); err == badger.ErrTxnTooBig {
+			err = txn.Commit()
+			if err != nil {
+				return err
+			}
+			txn = e.Db.NewTransaction(true)
+			err = txn.SetEntry(entry)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	_ = misc.RetryWith(context.TODO(), e.queryTimeout, e.retries, setBadgerEntry)
+	_ = txn.Commit()
```

How come we're not returning errors here?

---

We can throw an error and handle it at the parent function but live events is a functionality where we are okay with the server not panicking even if something goes wrong with the cache(unable to write to cache or read from cache) but I should probably add some error logs over here

---

You could return an error and let the parent log. I don't see a particular reason why a logger would be needed here.

---

Done